torchtune/models/phi3/_tokenizer.py

```diff
@@ -4,101 +4,81 @@
 # This source code is licensed under the BSD-style license found in the
 # LICENSE file in the root directory of this source tree.
 
-from typing import List, Optional, Tuple
+from typing import Dict, List, Optional, Tuple
 
-from sentencepiece import SentencePieceProcessor
 from torchtune.data import Message, truncate
-
-
-class Phi3MiniSentencePieceTokenizer:
-    """A wrapper around SentencePieceProcessor.
+from torchtune.modules.tokenizers import ModelTokenizer, SentencePieceBaseTokenizer
+
+PHI3_SPECIAL_TOKENS = {
+    "<|endoftext|>": 32000,
+    "<|assistant|>": 32001,
+    "<|placeholder1|>": 32002,
+    "<|placeholder2|>": 32003,
+    "<|placeholder3|>": 32004,
+    "<|placeholder4|>": 32005,
+    "<|system|>": 32006,
+    "<|end|>": 32007,
+    "<|placeholder5|>": 32008,
+    "<|placeholder6|>": 32009,
+    "<|user|>": 32010,
+}
+
+
+class Phi3MiniTokenizer(ModelTokenizer):
+    """
+    SentencePiece tokenizer configured with Phi3 Mini's special tokens.
 
     Args:
         path (str): Path to pretrained tokenizer file.
+        special_tokens (Optional[Dict[str, int]]): mapping containing special text tokens and
+            their registered token IDs. If left as None, this will be set to the canonical
+            Phi3 special tokens.
 
     Example:
         # Accepts only non-batched input for now
-        >>> tokenizer = SentencePieceTokenizer("/path/to/spm_model")
-        >>> tokenized_text = SentencePieceTokenizer.encode("Hello world!", add_bos=True, add_eos=True)
+        >>> tokenizer = Phi3MiniTokenizer("/path/to/spm_model")
+        >>> tokenized_text = tokenizer.encode("Hello world!", add_bos=True, add_eos=True)
         >>> print(tokenized_text)
         [1, 31587, 29644, 102, 2]
     """
 
     def __init__(
         self,
         path: str,
+        special_tokens: Optional[Dict[str, int]] = None,
     ):
-        spm_model = SentencePieceProcessor()
-        spm_model.load(path)
-        self.spm_model = spm_model
-
-        self.special_tokens = {
-            "<|endoftext|>": 32000,
-            "<|assistant|>": 32001,
-            "<|placeholder1|>": 32002,
-            "<|placeholder2|>": 32003,
-            "<|placeholder3|>": 32004,
-            "<|placeholder4|>": 32005,
-            "<|system|>": 32006,
-            "<|end|>": 32007,
-            "<|placeholder5|>": 32008,
-            "<|placeholder6|>": 32009,
-            "<|user|>": 32010,
-        }
-
-        self.vocab_size = spm_model.vocab_size()
-        self.bos_id = spm_model.bos_id()
+        self._spm_model = SentencePieceBaseTokenizer(path)
+
+        self.special_tokens = special_tokens or PHI3_SPECIAL_TOKENS
```

Comment:
nit: what if I want to explicitly pass an empty dict (i.e. no special tokens)? I think this will override that, right? (Maybe this isn't a realistic case, I'm not super familiar with phi3 tokenizer usage)

Reply:
It will override it, but I don't think this is a case we need to worry about. If there's no special tokens, then why use the Phi3Tokenizer? But, to be extra cautious, it's easy to just make this check for None explicitly.